Citylist scene using MVVM design pattern

vandermesis opened the pull request.

vandermesis committed:
Prepare MVVM CityList scene

vandermesis committed:
Setup CityListControllerMVVM and CityListCreatorMVVM

Setup:
	Setup CityListController xib
	Setup CityListController with tableview requirements
	Setup CityListCreator to inject ViewModel to ViewController

vandermesis committed:
Setup CityListViewModel with networking and repository

vandermesis committed:
Setup CityListMVVM scene with delegate design pattern

Changes:
	Setup CityListMVVMController with viewmodel delegate
	Setup CityListMVVMViewModel with networking and repository methods
	Use Delegate design pattern to pass data to viewcontroller

vandermesis committed:
Add CityListCellManagerMVVM

Changes:
	Move CityListController tableview setup to CellManager
	Inject CellManager to CityListController using Constructor Injection
	Move CityWeatherDisplayable to CellManager

vandermesis committed:
Refactor CityListViewModelMVVMDelegate method

vandermesis committed:
Some naming changes

SredzinskiPatryk commented on CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift:
```diff
@@ -0,0 +1,52 @@
+    func didSelectCityCell(city: CityWeather)
+}
+
+final class CityListCellManagerMVVM: NSObject {
```
Do you have to use nsobject?

vandermesis replied:
If I don't use it here then I get error
`Type 'CityListCellManagerMVVM' does not conform to protocol 'NSObjectProtocol'`
And to fix that I have to import 13 methods:
```
func isEqual(_ object: Any?) -> Bool {
        <#code#>
    }
    
    var hash: Int {
        <#code#>
    }
    
    var superclass: AnyClass? {
        <#code#>
    }

    func `self`() -> Self {
        <#code#>
    }

    func perform(_ aSelector: Selector!) -> Unmanaged<AnyObject>! {
        <#code#>
    }

    func perform(_ aSelector: Selector!, with object: Any!) -> Unmanaged<AnyObject>! {
        <#code#>
    }

    func perform(_ aSelector: Selector!, with object1: Any!, with object2: Any!) -> Unmanaged<AnyObject>! {
        <#code#>
    }

    func isProxy() -> Bool {
        <#code#>
    }

    func isKind(of aClass: AnyClass) -> Bool {
        <#code#>
    }

    func isMember(of aClass: AnyClass) -> Bool {
        <#code#>
    }

    func conforms(to aProtocol: Protocol) -> Bool {
        <#code#>
    }

    func responds(to aSelector: Selector!) -> Bool {
        <#code#>
    }

    var description: String {
        <#code#>
    }
```

SredzinskiPatryk commented on CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift:
```diff
@@ -0,0 +1,52 @@
+    func setup(with citiesWeather: [CityWeather]) {
+        self.citiesWeather = citiesWeather
+    }
+}
```
In most of MVVM Modells cell manager is not exactly working like that.

SredzinskiPatryk replied:
It’s fine and we can leave it as it, but I will also tell you how it looks like in standard products. Już should keep your data souls in view model and Ask cell manager only for cells. Most of them would only have only one method called build cell

SredzinskiPatryk commented on CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift:
```diff
@@ -0,0 +1,92 @@
+
+    @IBOutlet private weak var cityListTableView: UITableView!
+
+    private var citiesWeatherDataSource = [CityWeather]()
```
Keep the data source in view model that’s the whole idea of MVVM

vandermesis replied:
I didn't notice that it is not needed here. That var is not used anywhere. I just forgot to delete it after refactoring. :/

vandermesis replied:
But I get a point about not using data source in controller but only in viewmodel. :) I didn't know that.

vandermesis committed:
Refactor CityListCellManagerMVVM

Changes:
	Move delegate and datasource methods back to controller
	Add spinner to CityListControllerMVVM

vandermesis committed:
Merge remote-tracking branch 'refs/remotes/origin/master'

Conflicts:
	CleanWeather.xcodeproj/project.pbxproj
	CleanWeather/Scenes/CityList/CityListController.xib

vandermesis committed:
Add missing files after resolve merge conflicts

SredzinskiPatryk commented on CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift:
```diff
@@ -0,0 +1,34 @@
+
+final class CityListCellManagerMVVM: NSObject {
+
+    var citiesWeather = [CityWeather]()
```
get rid of this as well, pass array and index in build Cell Methods

SredzinskiPatryk commented on CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift:
```diff
@@ -0,0 +1,34 @@
+        return cell
+    }
+
+    func didSelectCityCell(indexPath: IndexPath) -> CityWeather {
```
you can keep it only in controller, no need to continue asking for cell manager in controllers tableview delegate

SredzinskiPatryk commented on CleanWeather/Scenes/CityListMVVM/CityListCellManagerMVVM.swift:
```diff
@@ -0,0 +1,34 @@
+        return citiesWeather[indexPath.row]
+    }
+
+    func setupCellManager(with citiesWeather: [CityWeather]) {
```
no need this

SredzinskiPatryk commented on CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift:
```diff
@@ -0,0 +1,103 @@
+extension CityListControllerMVVM: UITableViewDataSource {
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        cellManager.citiesWeather.count
```
viewModel.citiesWeather.count

SredzinskiPatryk commented on CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift:
```diff
@@ -0,0 +1,103 @@
+
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let city = cellManager.didSelectCityCell(indexPath: indexPath)
+        navigateToCityForecast(city: city)
```
let city = viewModel.citiesWeather[indexPath.row]

SredzinskiPatryk replied:
no need cell manager

SredzinskiPatryk commented on CleanWeather/Scenes/CityListMVVM/CityListControllerMVVM.swift:
```diff
@@ -0,0 +1,103 @@
+
+    func didUpdateFavouriteCitiesWeather(citiesWeather: [CityWeather]) {
+        toogleSpinner(false)
+        cellManager.setupCellManager(with: citiesWeather)
```
no need this

vandermesis committed:
Update README.md

vandermesis committed:
Fix CellManager and move datasource to viewmodel only

SredzinskiPatryk commented on CleanWeather/Scenes/CityListMVVM/CityListViewModelMVVM.swift:
```diff
@@ -0,0 +1,110 @@
+        repository.fetchFavouriteCities { result in
+            switch result {
+            case .success(let cities):
+                self.fetchCitiesWeather(cities: cities) { result in
```
put it into separated method

SredzinskiPatryk commented on CleanWeather/Scenes/CityListMVVM/CityListViewModelMVVM.swift:
```diff
@@ -0,0 +1,110 @@
+        case .failure(let error):
+            return .failure(error)
+        case .success(let apiResponse):
+            guard let temp = apiResponse.currently.temperature else {
```
put it into separated method

vandermesis committed:
Refactor CityListViewModelMVVM

Changes:
	Split getFavouriteCities method
	Split convertToCityWeatherResult method
	Change methods order to reflect methods flow
